File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.py
class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        if source==destination:
            return True

        s = dict()

        # save all the edges into a dict
        for edge1, edge2 in edges:
            if edge1 not in s:
                s[edge1] = [edge2]
            else:
                s[edge1].append(edge2)
            if edge2 not in s:
                s[edge2] = [edge1]
            else:
                s[edge2].append(edge1)

        # go through all the possible paths, dfs
        visited = set()
        q = deque()
        for x in s[source]:
            q.append(x)
        while q:
            current = q.pop()
            # if the node is visited before, skip
            if current in visited:
                continue
            elif destination == current:
                return True
            else:
                visited.add(current)
                for x in s[current]:
                    if x not in visited:
                        q.append(x)

        return False

File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.cpp
class Solution {
public:
    bool validPath(int n, vector<vector<int>>& edges, int source, int destination) {
        if (source == destination) return true;
        // Store all valid paths into mp
        unordered_map<int, vector<int>> mp;
        for (int i=0; i<edges.size(); ++i) {
            // because all paths are bi-directional
            mp[edges[i][0]].push_back(edges[i][1]);
            mp[edges[i][1]].push_back(edges[i][0]);
        }

        unordered_set<int> visited;
        queue<int> q;
        q.push(source);
        visited.insert(source);

        while (!q.empty()) {
            int nxt = q.front();
            q.pop();

            if (nxt == destination)
                return true;

            for (auto it=mp[nxt].begin(); it<mp[nxt].end(); ++it) {
                // if not visited before
                if (visited.find(*it) == visited.end()) {
                    q.push(*it);
                    visited.insert(*it);
                }
            }
        }

        return false;
    }
};